pkg/controller/volume/expand/cache/volume_resize_map.go

```diff
@@ -0,0 +1,303 @@
+// performing file system resize
+type PvcWithResizeRequest struct {
+	// PVC that needs to be resized
+	PVC *v1.PersistentVolumeClaim
+	// Volume spec object that can be used from volume plugins
+	VolumeSpec *volume.Spec
+	// Current volume size
+	CurrentSize resource.Quantity
+	// Expended volume size
+	ExpectedSize resource.Quantity
+}
+
+// UniquePvcKey returns unique key of the PVC based on its UID
+func (pvcr *PvcWithResizeRequest) UniquePvcKey() types.UniquePvcName {
+	return types.UniquePvcName(pvcr.PVC.UID)
+}
+
+// QualifiedName returns namespace and name combination of the PVC
+func (pvcr *PvcWithResizeRequest) QualifiedName() string {
+	return strings.JoinQualifiedName(pvcr.PVC.Namespace, pvcr.PVC.Name)
+}
+
+// NewVolumeResizeMap returns new VolumeResizeMap which acts as a cache
+// for holding pending resize requests.
+func NewVolumeResizeMap(kubeClient clientset.Interface) VolumeResizeMap {
+	resizeMap := &volumeResizeMap{}
+	resizeMap.pvcrs = make(map[types.UniquePvcName]*PvcWithResizeRequest)
+	resizeMap.kubeClient = kubeClient
+	return resizeMap
+}
+
+func (resizeMap *volumeResizeMap) AddPVCUpdate(newPvc *v1.PersistentVolumeClaim, oldPvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+	if newPvc.Namespace != pv.Spec.ClaimRef.Namespace || newPvc.Name != pv.Spec.ClaimRef.Name {
+		glog.Errorf("Persistent Volume is not mapped to PVC being updated : %s", util.ClaimToClaimKey(newPvc))
+		return
+	}
+
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	newSize := newPvc.Spec.Resources.Requests[v1.ResourceStorage]
+	oldSize := oldPvc.Spec.Resources.Requests[v1.ResourceStorage]
+
+	if newSize.Cmp(oldSize) > 0 {
+		pvcRequest := &PvcWithResizeRequest{
+			PVC:          newPvc,
+			CurrentSize:  newPvc.Status.Capacity[v1.ResourceStorage],
+			ExpectedSize: newSize,
+			VolumeSpec:   volume.NewSpecFromPersistentVolume(pv, false),
+		}
+		resizeMap.pvcrs[types.UniquePvcName(newPvc.UID)] = pvcRequest
+	}
+}
+
+// VerifyAndSoftAddPVC adds PVC for resizing only if there is no
+// resize request pending and if PV size doesn't satisfy pvc size
+func (resizeMap *volumeResizeMap) VerifyAndSoftAddPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+	if pvc.Namespace != pv.Spec.ClaimRef.Namespace || pvc.Name != pv.Spec.ClaimRef.Name {
+		glog.Errorf("Persistent Volume is not mapped to PVC being updated : %s", util.ClaimToClaimKey(pvc))
+		return
+	}
+
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	pvcUniqueName := types.UniquePvcName(pvc.UID)
+
+	if _, ok := resizeMap.pvcrs[pvcUniqueName]; !ok {
+		pvcSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+		pvSize := pv.Spec.Capacity[v1.ResourceStorage]
+		// If PV size is less than pvc size. The reason here we compare with pv size
+		// rather than pvc.Status.Size is because if a resize is already in progress
+		// then pv size would reflect new value and if pv size is more than pvc
+		// size then no resize will be necessary.
+		if pvSize.Cmp(pvcSize) < 0 {
```

Comment:
You should compare also pvc.status.capacity with pvc.spec.capacity. For example, when the controller updates pv.spec.capacity with new value and crashes just before saving pvc.status.capacity then there is no path that would make sure that something updates the status.

Again, the best would be check *only* pvc.spec vs pvc.status and the operation itself would check what needs to be done - update pvc status conditions, resize PV, update pv.status, update pvc.status. There is some overhead when the controller is waiting for FS resize, but that's IMO it's negligible.

Reply:
I have thought long and hard about it and comparing `pvc.spec.capacity` and `pvc.status.capacity` here - will result in double resizing of every volume type that requires file system resize.  IMO it is unnecessary churn and optimizing for the worst 2% of case (namely controller crash). 

Is there a way, we can handle this and avoid double resizing of same volume?  I am thinking that, rather than solving this ourselves - may be we should provide admin mechanism to sort it out. namely:

1. We can log an event if we notice that a pvc's status and spec capacity don't match and is in `ResizeStarted` state for too long..
2. If a pvc's status and spec doesn't match and is not in `ResizeInProgress`, then it gets added to resize map.

So basically what we are saying is, admin can manually edit the PVC and drop `ResizeInProgress` field from pvc.Status and that would make populator pick it up again.

Reply:
There is another problem with solely relying on `pvc.Spec.Capacity` and `pvc.Status.Capacity`. When you create a new PVC and when it gets dynamically provisioned - I am seeing a state where `pvc.Spec.Capacity` has user requested size but `pvc.Status.Capacity` is zero.  It is temporary and it does seem to get fixed in next update. But relying only on `pvc.Spec.Capacity` and `pvc.Status.Capacity` would result in expand controller trying to resize such a volume.

Reply:
> When you create a new PVC and when it gets dynamically provisioned - I am seeing a state where pvc.Spec.Capacity has user requested size but pvc.Status.Capacity is zero.

Is the PVC bound at that time? It's error if it's reported as Bound and status.capacity does not reflect the PV size.

Reply:
> will result in double resizing of every volume type that requires file system resize

```golang

func resizeOperation(pv, pvc) {
    if pvc.spec.Capacity <= pvc.status.Capacity {
        return
    }

    if !hasResizingCondition(pvc.status) {
        // Add the condition if it's not there yet
        addResizingCondition(pvc)
    }

    if pv.spec.capacity < pvc.spec.capacity {
        // PV has not been resized yet, resize it
        newSize = plugin.ExpandVolumeDevice(pvc.spec.capacity)
        saveNewPVSize(pv, newSize)
    }

    if needFSResize() {
        // waiting for FS resize, it will update pvc.status.capacity and remove the condition in the end
        return
    } else {
        saveNewPVCSizeAndRemoveResizingCondition(pvc, newSize)
    }
}
```

This should be idempotent and can be interrupted at any time without any loss. It extends the volume twice only when it crashes between `plugin.ExpandVolumeDevice` and `saveNewPVSize`, which is IMO rare enough. If anything returns error, the operation can be restarted and it will do only the necessary changes.

When a PV waits for a FS resize, this operation just checks that a condition is set, compares some sizes and returns.

Reply:
ack done